substrate/client/network/src/litep2p/shim/notification/peerset.rs

```diff
@@ -0,0 +1,1276 @@
+				log::debug!(target: LOG_TARGET, "{}: reserved peer {peer:?} disconnected", self.protocol);
+			},
+			state => {
+				log::warn!(target: LOG_TARGET, "{}: invalid state for disconnected peer {peer:?}: {state:?}", self.protocol);
+				debug_assert!(false);
+			},
+		}
+		*state = PeerState::Backoff;
+
+		self.connected_peers.fetch_sub(1usize, Ordering::Relaxed);
+		self.pending_backoffs.push(Box::pin(async move {
+			Delay::new(DEFAULT_BACKOFF).await;
+			(peer, DISCONNECT_ADJUSTMENT)
+		}));
+	}
+
+	/// Report to [`Peerset`] that an inbound substream was opened and that it should validate it.
+	pub fn report_inbound_substream(&mut self, peer: PeerId) -> ValidationResult {
+		log::trace!(target: LOG_TARGET, "{}: inbound substream from {peer:?}", self.protocol);
+
+		let state = self.peers.entry(peer).or_insert(PeerState::Disconnected);
+		let reserved_peer = self.reserved_peers.contains(&peer);
+
+		match state {
+			// disconnected peers proceed directly to inbound slot allocation
+			PeerState::Disconnected => {},
+			// peer is backed off but if it can be accepted (either a reserved peer or inbound slot
+			// available), accept the peer and then just ignore the back-off timer when it expires
+			PeerState::Backoff => {
+				if !self.reserved_peers.contains(&peer) && self.num_in == self.max_in {
+					log::trace!(
+						target: LOG_TARGET,
+						"{}: ({peer:?}) is backed-off and cannot accept, reject inbound substream",
+						self.protocol,
+					);
+
+					return ValidationResult::Reject
+				}
+			},
+			// `Peerset` had initiated an outbound substream but litep2p had received an inbound
+			// substream before the command was received. As litep2p is the source of truth as far
+			// as substream states go, `Peerset` must update its own state and attempt to accept the
+			// peer as inbound if there are enough slots.
+			PeerState::Opening { direction: Direction::Outbound(reserved) } => {
+				log::trace!(
+					target: LOG_TARGET,
+					"{}: inbound substream received for {peer:?} that was marked outbound",
+					self.protocol,
+				);
+
+				match (reserved_peer, &reserved) {
+					(true, &Reserved::Yes) => {},
+					(false, &Reserved::No) => {},
+					_ => panic!(
+						"{}: state mismatch for {peer:?}, {reserved_peer} {reserved:?}",
+						self.protocol
+					),
+				}
+
+				// since the peer was not a reserved peer when the outbound substream was opened and
+				// there are no free inbound slots, adjust outbound slot count and reject the peer
+				if std::matches!(reserved, Reserved::No) {
+					adjust_or_warn!(
+						self.num_out,
+						self.protocol,
+						peer,
+						Direction::Outbound(Reserved::No)
+					);
+
+					if self.num_in >= self.max_in {
+						log::debug!(
+							target: LOG_TARGET,
+							"{}: inbound substream for {peer:?} cannot be accepted because there aren't any free slots",
+							self.protocol,
+						);
+
+						*state = PeerState::Disconnected;
+						return ValidationResult::Reject
```

Comment:
Is it possible that `litep2p` is notified about an outbound connection via `handle.open_substream_batch()` for this `Opening` peer actually after we put the peer into `Disconnected` state? I.e., we could have more `litep2p` connections than `Peerset` thinks there are?

Reply:
Actually, it would help to have the state diagrams of `Peerset` and litep2p notification protocol displayed alongside each other. Since there is also `NotificationProtocol` shim between, it's a little bit difficult to understand the state synchronization.

Reply:
Yes good catch. I actually run into this in Versi today. BEEFY opened an outbound substream but at the same time there was an inbound substream received by litep2p but the streams were completely uncorrelated, BEEFY tried to convert the outbound substream to an inbound substream and failed because there weren't enough slots and then litep2p processed the outbound substream request separately which lead to an "invalid" state transition in `Peerset`. I'll have to modify this code a bit.

I'll draw some diagrams after the Versi tests are done and when there aren't any new issues that I can find

Reply:
I've simplified the logic a bit. If `Peerset` has opened an outbound substream and an inbound substream is received while the outbound is opening, the substream is accepted but the slot count is not updated and the peer is kept as an outbound peer. This sidesteps the issue where there aren't enough inbound slots, the peer is rejected but an outbound substream is opened nonetheless. If the connection opens successfully, the direction that `Peerset` considers to be correct is reported to protocol.

The direction is only used by `SyncingEngine` to make sure it won't accept excess inbound substreams because of a design limitation in how light peers are represented in the inbound slot count. That we can refactor in the future and get rid of this workaround.

Ref: https://github.com/altonen/litep2p/issues/34

Reply:
Makes sense, this should work out.